scripts/deleteAndLoadSnapsets.js

```diff
@@ -0,0 +1,398 @@
+/*!
+Copyright 2018 OCAD University
+
+Licensed under the New BSD license. You may not use this file except in
+compliance with this License.
+
+You may obtain a copy of the License at
+https://github.com/GPII/universal/blob/master/LICENSE.txt
+*/
+
+// This script modifies the preferences database:
+// 1. Retrieves all the Prefs Safes of type "snapset" (prefsSafesType = "snapset") from the databsse.
+// 2. Retrieves all the GPII Keys associated with each snapset Prefs Safe so found,
+// 3. Deletes these Prefs Safes and their associated GPII Keys from the database,
+// 4. Uploads the new Prefs Safes and their GPII Keys to the database,
+// 5. Uploads demo user Prefs Safes and their Keys in to the database, if they are not already present.
+// A sample command that runs this script:
+// node deleteAndLoadSnapsets.js $COUCHDBURL $BUILD_DATA_DIR $BUILD_DEMOUSER_DIR
+
+"use strict";
+
+var http = require("http"),
+    url = require("url"),
+    fs = require("fs"),
+    fluid = require("infusion");
+
+var gpii = fluid.registerNamespace("gpii");
+fluid.registerNamespace("gpii.dataLoader");
+fluid.setLogging(fluid.logLevel.INFO);
+
+// Handle command line
+if (process.argv.length < 5) {
+    fluid.log("Usage: node deleteAndLoadSnapsets.js $COUCHDB_URL $BUILD_DATA_DIR $BUILD_DEMOUSER_DIR [--justDelete]");
+    process.exit(1);
+}
+
+var dbLoader = gpii.dataLoader;
+
+/**
+ * Create a set of options for data loader and a function to retreive them.
+ * The options are based on the command line parameters and a set of database
+ * constants.
+ * @param {Array} processArgv - The command line arguments.
+ */
+dbLoader.initOptions = function (processArgv) {
+    var dbOptions = {};
+    dbOptions.couchDbUrl = processArgv[2];
+    dbOptions.buildDataDir = processArgv[3];
+    dbOptions.demoUserDir = processArgv[4];
+    if (processArgv.length > 5 && processArgv[5] === "--justDelete") { // for debugging.
+        dbOptions.justDelete = true;
+    } else {
+        dbOptions.justDelete = false;
+    }
+
+    // Set up database specific options
+    dbOptions.prefsSafesViewUrl = dbOptions.couchDbUrl + "/_design/views/_view/findSnapsetPrefsSafes";
+    dbOptions.gpiiKeysViewUrl = dbOptions.couchDbUrl + "/_design/views/_view/findAllGpiiKeys";
+    dbOptions.parsedCouchDbUrl = url.parse(dbOptions.couchDbUrl);
+    dbOptions.snapsetPrefsSafes = [];
+    dbOptions.gpiiKeys = [];
+    dbOptions.postOptions = {
+        hostname: dbOptions.parsedCouchDbUrl.hostname,
+        port: dbOptions.parsedCouchDbUrl.port,
+        path: "/gpii/_bulk_docs",
+        method: "POST",
+        headers: {
+            "Accept": "application/json",
+            "Content-Length": 0,          // IMPORTANT: FILL IN PER REQUEST
+            "Content-Type": "application/json"
+        }
+    };
+
+    // Set up function to get at these options.
+    gpii.dataLoader.getOptions = function () {
```

Comment:
Still a pretty bizarre approach of writing a stateful global function. Please just end this function with "return dbOptions", and then pass this as an argument to dbLoader.processSnapsets.

Reply:
I misread [your comment](https://github.com/GPII/universal/pull/626#discussion_r206926250) about this as: put these globals into a local options variable and provide an accessor for them.  But you said nothing about an accessor; and, it's unneccessary, as you point out.  I've removed it.